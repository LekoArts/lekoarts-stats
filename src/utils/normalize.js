import _ from 'lodash'

export function flatten(data) {
  return data.flatMap((g) => {
    return g.repos.map((r) => {
      return {
        name: r.name,
        datetime: g.datetime,
        stars: r.stars,
        forks: r.forks,
        url: r.url,
      }
    })
  })
}

/**
 * _.mapValues: https://lodash.com/docs/4.17.15#mapValues
 * _.groupBy: https://lodash.com/docs/4.17.15#groupBy
 * _.omit: https://lodash.com/docs/4.17.15#omit
 */

export function constructShape(flat, name, omit) {
  return _.mapValues(_.groupBy(flat, name), (list) => list.map((entry) => _.omit(entry, omit)))
}

export function normalizeGithub(data, name, omit) {
  const flat = flatten(data)
  return constructShape(flat, name, omit)
}

export function nivoGithubFormatter(data, name) {
  const keys = Object.keys(data)
  const nivoData = []

  for (const key of keys) {
    const entry = data[key]

    const values = entry.map((e) => ({
      x: e.datetime,
      y: e[name],
    }))

    const obj = {
      id: key,
      data: values,
    }

    nivoData.push(obj)
  }

  return nivoData.reverse()
}

export function nivoTwitterFormatter(data, name) {
  const values = data.map((e) => ({
    x: e.datetime,
    y: e[name],
  }))

  return [
    {
      id: name,
      data: values,
    },
  ]
}
