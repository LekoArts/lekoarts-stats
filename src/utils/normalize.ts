import { groupBy, mapValues, omit } from 'lodash-es'

/**
 * _.mapValues: https://lodash.com/docs/4.17.15#mapValues
 * _.groupBy: https://lodash.com/docs/4.17.15#groupBy
 * _.omit: https://lodash.com/docs/4.17.15#omit
 */

/**
 * Formats 2020-07-18 14:02:07.078+00 to 2020-07-18
 */
export function formatCreatedAt(date: string) {
  return date.split(' ')[0]
}

interface IGitHubEntry {
  id: number
  name: string
  createdAt: string
  forks: number
  stars: number
}

export function normalizeGithub(data: IGitHubEntry[], name: keyof IGitHubEntry) {
  return mapValues(groupBy(data, name), list => list.map(entry => omit(entry, name))) as GitHubFormatterDataInput
}

interface GitHubFormatterDataInput {
  [key: string]: Omit<IGitHubEntry, 'name'>[]
}

export function nivoGithubFormatter(data: GitHubFormatterDataInput, name: keyof Omit<IGitHubEntry, 'name'>) {
  const keys = Object.keys(data)
  const nivoData = []

  for (const key of keys) {
    const entry = data[key]

    const values = entry.map(e => ({
      x: formatCreatedAt(e.createdAt),
      y: e[name],
    }))

    const obj = {
      id: key,
      data: values,
    }

    nivoData.push(obj)
  }

  return nivoData.reverse()
}

interface ITwitterEntry {
  id: number
  createdAt: string
  tweets: number
  followers: number
}

export function nivoTwitterFormatter(data: ITwitterEntry[]) {
  const followersValues = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e.followers,
  }))
  const tweetsValues = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e.tweets,
  }))

  return [
    {
      id: 'followers',
      data: followersValues,
    },
    {
      id: 'tweets',
      data: tweetsValues,
    },
  ]
}

interface IMastodonEntry {
  id: number
  createdAt: string
  followersCount: number
  tootsCount: number
}

export function nivoMastodonFormatter(data: IMastodonEntry[]) {
  const followersValues = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e.followersCount,
  }))
  const tweetsValues = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e.tootsCount,
  }))

  return [
    {
      id: 'followers',
      data: followersValues,
    },
    {
      id: 'toots',
      data: tweetsValues,
    },
  ]
}

interface ITraktEntry {
  id: number
  createdAt: string
  moviesWatched: number
  showsWatched: number
  episodesWatched: number
  ratings: number
}

export function nivoTraktFormatter(data: ITraktEntry[], name: keyof ITraktEntry) {
  const values = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e[name],
  }))

  return [
    {
      id: name,
      data: values,
    },
  ]
}
