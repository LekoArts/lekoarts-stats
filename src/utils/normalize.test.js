import { normalizeGithub, flatten, constructShape, nivoGithubFormatter, nivoTwitterFormatter } from './normalize'
import { data as githubData } from './__fixtures__/github'
import { data as twitterData } from './__fixtures__/twitter'

describe('flatten', () => {
  it('should create new object shape', () => {
    const flat = flatten(githubData)

    const objectShape = [
      {
        datetime: '2020-07-17',
        forks: 154,
        name: 'gatsby-starter-minimal-blog',
        stars: 517,
        url: 'https://github.com/LekoArts/gatsby-starter-minimal-blog',
      },
    ]

    expect(flat).toMatchSnapshot()
    expect(flat).toEqual(expect.arrayContaining(objectShape))
    expect(flat[0].id).not.toBeDefined()
  })
  it('should create entries in correct order', () => {
    const flat = flatten(githubData)
    const length = flat.length

    expect(flat[0]).toStrictEqual({
      datetime: '2020-07-17',
      forks: 243,
      name: 'gatsby-themes',
      stars: 745,
      url: 'https://github.com/LekoArts/gatsby-themes',
    })
    expect(flat[length - 1]).toStrictEqual({
      datetime: '2020-07-18',
      forks: 18,
      name: 'gatsby-starter-portfolio-bella',
      stars: 113,
      url: 'https://github.com/LekoArts/gatsby-starter-portfolio-bella',
    })
  })
})

describe('constructShape', () => {
  it('should group data by name and omit name', () => {
    const flat = flatten(githubData)
    const shape = constructShape(flat, 'name', 'name')
    const namedEntry = shape['gatsby-starter-minimal-blog']

    expect(shape).toMatchSnapshot()
    expect(namedEntry).toStrictEqual([
      {
        datetime: '2020-07-17',
        forks: 154,
        stars: 517,
        url: 'https://github.com/LekoArts/gatsby-starter-minimal-blog',
      },
      {
        datetime: '2020-07-18',
        forks: 153,
        stars: 517,
        url: 'https://github.com/LekoArts/gatsby-starter-minimal-blog',
      },
    ])
  })
})

describe('normalizeGithub', () => {
  it('should flatten and group data by name', () => {
    const normalized = normalizeGithub(githubData, 'name', 'name')
    const namedEntry = normalized['gatsby-starter-minimal-blog']

    expect(namedEntry).toStrictEqual([
      {
        datetime: '2020-07-17',
        forks: 154,
        stars: 517,
        url: 'https://github.com/LekoArts/gatsby-starter-minimal-blog',
      },
      {
        datetime: '2020-07-18',
        forks: 153,
        stars: 517,
        url: 'https://github.com/LekoArts/gatsby-starter-minimal-blog',
      },
    ])
  })
})

describe('nivoGithubFormatter', () => {
  it('should output correct shape via input', () => {
    const normalized = normalizeGithub(githubData, 'name', 'name')
    const nivo = nivoGithubFormatter(normalized, 'stars')

    const objectShape = [
      {
        data: [
          {
            x: '2020-07-17',
            y: 517,
          },
          {
            x: '2020-07-18',
            y: 517,
          },
        ],
        id: 'gatsby-starter-minimal-blog',
      },
    ]

    expect(nivo).toMatchSnapshot()
    expect(nivo).toEqual(expect.arrayContaining(objectShape))
  })
})

describe('nivoTwitterFormatter', () => {
  it('should output correct shape via input', () => {
    const nivo = nivoTwitterFormatter(twitterData, 'followers')

    const objectShape = [
      {
        data: [
          {
            x: '2020-07-16',
            y: 1126,
          },
          {
            x: '2020-07-17',
            y: 1137,
          },
          {
            x: '2020-07-18',
            y: 1150,
          },
        ],
        id: 'followers',
      },
    ]

    expect(nivo).toMatchSnapshot()
    expect(nivo).toEqual(expect.arrayContaining(objectShape))
  })
})
