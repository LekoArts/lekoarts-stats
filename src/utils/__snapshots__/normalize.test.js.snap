// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructShape should group data by name and omit name 1`] = `
Object {
  "gatsby-starter-minimal-blog": Array [
    Object {
      "datetime": "2020-07-17",
      "forks": 154,
      "stars": 517,
      "url": "https://github.com/LekoArts/gatsby-starter-minimal-blog",
    },
    Object {
      "datetime": "2020-07-18",
      "forks": 153,
      "stars": 517,
      "url": "https://github.com/LekoArts/gatsby-starter-minimal-blog",
    },
  ],
  "gatsby-starter-portfolio-bella": Array [
    Object {
      "datetime": "2020-07-17",
      "forks": 19,
      "stars": 113,
      "url": "https://github.com/LekoArts/gatsby-starter-portfolio-bella",
    },
    Object {
      "datetime": "2020-07-18",
      "forks": 18,
      "stars": 113,
      "url": "https://github.com/LekoArts/gatsby-starter-portfolio-bella",
    },
  ],
  "gatsby-themes": Array [
    Object {
      "datetime": "2020-07-17",
      "forks": 243,
      "stars": 745,
      "url": "https://github.com/LekoArts/gatsby-themes",
    },
    Object {
      "datetime": "2020-07-18",
      "forks": 245,
      "stars": 749,
      "url": "https://github.com/LekoArts/gatsby-themes",
    },
  ],
}
`;

exports[`flatten should create new object shape 1`] = `
Array [
  Object {
    "datetime": "2020-07-17",
    "forks": 243,
    "name": "gatsby-themes",
    "stars": 745,
    "url": "https://github.com/LekoArts/gatsby-themes",
  },
  Object {
    "datetime": "2020-07-17",
    "forks": 154,
    "name": "gatsby-starter-minimal-blog",
    "stars": 517,
    "url": "https://github.com/LekoArts/gatsby-starter-minimal-blog",
  },
  Object {
    "datetime": "2020-07-17",
    "forks": 19,
    "name": "gatsby-starter-portfolio-bella",
    "stars": 113,
    "url": "https://github.com/LekoArts/gatsby-starter-portfolio-bella",
  },
  Object {
    "datetime": "2020-07-18",
    "forks": 245,
    "name": "gatsby-themes",
    "stars": 749,
    "url": "https://github.com/LekoArts/gatsby-themes",
  },
  Object {
    "datetime": "2020-07-18",
    "forks": 153,
    "name": "gatsby-starter-minimal-blog",
    "stars": 517,
    "url": "https://github.com/LekoArts/gatsby-starter-minimal-blog",
  },
  Object {
    "datetime": "2020-07-18",
    "forks": 18,
    "name": "gatsby-starter-portfolio-bella",
    "stars": 113,
    "url": "https://github.com/LekoArts/gatsby-starter-portfolio-bella",
  },
]
`;

exports[`nivoGithubFormatter should output correct shape via input 1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "x": "2020-07-17",
        "y": 113,
      },
      Object {
        "x": "2020-07-18",
        "y": 113,
      },
    ],
    "id": "gatsby-starter-portfolio-bella",
  },
  Object {
    "data": Array [
      Object {
        "x": "2020-07-17",
        "y": 517,
      },
      Object {
        "x": "2020-07-18",
        "y": 517,
      },
    ],
    "id": "gatsby-starter-minimal-blog",
  },
  Object {
    "data": Array [
      Object {
        "x": "2020-07-17",
        "y": 745,
      },
      Object {
        "x": "2020-07-18",
        "y": 749,
      },
    ],
    "id": "gatsby-themes",
  },
]
`;

exports[`nivoTwitterFormatter should output correct shape via input 1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "x": "2020-07-16",
        "y": 1126,
      },
      Object {
        "x": "2020-07-17",
        "y": 1137,
      },
      Object {
        "x": "2020-07-18",
        "y": 1150,
      },
    ],
    "id": "followers",
  },
]
`;
